File: e-learn-backend/ml-service/chatbot/chatbot.py
import sys
import openai
import os

# Set your OpenAI API key (ensure OPENAI_API_KEY is set in environment)
openai.api_key = os.getenv("OPENAI_API_KEY")

def get_chatbot_response(message):
    """Generate a response from the chatbot using OpenAI API."""
    try:
        response = openai.ChatCompletion.create(
            model="gpt-3.5-turbo",  # or "gpt-4" if available
            messages=[{"role": "user", "content": message}],
            max_tokens=50
        )
        return response.choices[0].message.content.strip()
    except Exception as e:
        print(f"Error generating chatbot response: {e}", file=sys.stderr)
        return "Sorry, I couldn't process your request."

if __name__ == "__main__":
    # Read the message from command-line arguments
    if len(sys.argv) > 1:
        user_message = sys.argv[1]
        print(f"Received message: {user_message}", file=sys.stderr)
        print(get_chatbot_response(user_message))
    else:
        print("Error: No message provided", file=sys.stderr)

File: e-learn-backend/routes/chatbot.js
const express = require("express");
const router = express.Router();
const axios = require("axios");
const Chatbot = require("../models/chatbotModel"); // Assuming you have this model for saving history

router.post("/", async (req, res) => {
  const userMessage = req.body.message;

  if (!userMessage || typeof userMessage !== "string" || !userMessage.trim()) {
    return res.status(400).json({ error: "Message is required" });
  }

  try {
    // ⭐ CHANGE 1: Change API endpoint to /v1/chat ⭐
    const cohereRes = await axios.post(
      "https://api.cohere.ai/v1/chat", // Changed from /v1/generate
      {
        model: "command-nightly",
        // ⭐ CHANGE 2: Use 'message' and 'preamble' instead of 'prompt' ⭐
        // The 'chat' endpoint uses a 'message' field for the current user input.
        message: userMessage,
        // 'preamble' is a good way to set the chatbot's persona (like "as a teacher").
        preamble: "You are a helpful and knowledgeable teacher. Always answer questions in a clear and educational manner.",
        max_tokens: 300,
        temperature: 0.7
        // For multi-turn conversations, you would also pass a 'chat_history' array here.
        // chat_history: [{ role: "USER", message: "Hello" }, { role: "CHATBOT", message: "Hi!" }],
      },
      {
        headers: {
          Authorization: `Bearer ${process.env.COHERE_API_KEY}`,
          "Content-Type": "application/json"
        }
      }
    );

    // ⭐ CHANGE 3: Adjust how to extract the bot response ⭐
    // For the /v1/chat endpoint, the main generated text is directly in 'cohereRes.data.text'
    const botResponse = cohereRes.data.text?.trim() || "Sorry, I couldn't generate a response.";

    // Save the conversation history (assuming your Chatbot model supports this)
    await Chatbot.create({ userMessage, botResponse });

    res.json({ response: botResponse });
  } catch (error) {
    console.error("❌ Cohere API Error:", error?.response?.data || error.message);
    const errorMessage = error?.response?.data?.message || "Failed to get a response from Cohere";
    res.status(500).json({ error: errorMessage });
  }
});

module.exports = router;